Added stub of min gcc/clang reqirements to CUDA 12.1

JulianVolodia opened the pull request:
when patched pytorch-nightly with dirty workaround, adding in version.py materialized version of file (not template, but from site-packages) setting cuda_version = 12.1 (previously: None) new error arised -- above problem is described here in issue of pytorch:
 https://github.com/pytorch/pytorch/issues/101801

It arised when using oobabooga/one-click-installers for Windows

Short:
```
File "G:\git-jv\ai\oobabooga_one-click-installersk-installers\installer_files\env\lib\site-packages\torch\__init__.py", line 91, in <module>
cuda_version_1 = cuda_version.replace('.', '_')
AttributeError: 'float' object has no attribute 'replace'
```

Long:
```
Requirement already satisfied: mpmath>=0.19 in g:\git-jv\ai\oobabooga_one-click-installersk-installers\installer_files\env\lib\site-packages (from sympy->torch>=1.4.0->accelerate==0.18.0->-r requirements.txt (line 5)) (1.2.1)
Installing collected packages: safetensors, transformers, accelerate, datasets
  Attempting uninstall: safetensors
    Found existing installation: safetensors 0.3.1
    Uninstalling safetensors-0.3.1:
      Successfully uninstalled safetensors-0.3.1
  Attempting uninstall: transformers
    Found existing installation: transformers 4.29.1
    Uninstalling transformers-4.29.1:
      Successfully uninstalled transformers-4.29.1
  Attempting uninstall: accelerate
    Found existing installation: accelerate 0.19.0
    Uninstalling accelerate-0.19.0:
      Successfully uninstalled accelerate-0.19.0
  Attempting uninstall: datasets
    Found existing installation: datasets 2.12.0
    Uninstalling datasets-2.12.0:
      Successfully uninstalled datasets-2.12.0
Successfully installed accelerate-0.18.0 datasets-2.10.1 safetensors-0.3.0 transformers-4.28.0
Processing g:\git-jv\ai\oobabooga_one-click-installersk-installers\text-generation-webui\repositories\gptq-for-llama
  Preparing metadata (setup.py): started
  Preparing metadata (setup.py): finished with status 'error'
  error: subprocess-exited-with-error

  python setup.py egg_info did not run successfully.
  exit code: 1

  [8 lines of output]
  Traceback (most recent call last):
    File "<string>", line 2, in <module>
    File "<pip-setuptools-caller>", line 34, in <module>
    File "G:\git-jv\ai\oobabooga_one-click-installersk-installers\text-generation-webui\repositories\GPTQ-for-LLaMa\setup.py", line 2, in <module>
      from torch.utils import cpp_extension
    File "G:\git-jv\ai\oobabooga_one-click-installersk-installers\installer_files\env\lib\site-packages\torch\__init__.py", line 91, in <module>
      cuda_version_1 = cuda_version.replace('.', '_')
  AttributeError: 'float' object has no attribute 'replace'
  [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

Encountered error while generating package metadata.

See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.
```

Related issue #101801 (but not fixes it)

JulianVolodia committed:
Wrapping cuda_version with str() in case it is float type

when patched pytorch-nightly with dirty workaround, adding in version.py
materialized version of file (not template, but from site-packages)
setting cuda_version = 12.1 (previously: None) new error arised
-- above problem is described here in issue of pytorch:
 https://github.com/pytorch/pytorch/issues/101801

It arised when using oobabooga/one-click-installers for Windows

Short:
```
File "G:\git-jv\ai\oobabooga_one-click-installersk-installers\installer_files\env\lib\site-packages\torch\__init__.py", line 91, in <module>
cuda_version_1 = cuda_version.replace('.', '_')
AttributeError: 'float' object has no attribute 'replace'
```

Long:
```
Requirement already satisfied: mpmath>=0.19 in g:\git-jv\ai\oobabooga_one-click-installersk-installers\installer_files\env\lib\site-packages (from sympy->torch>=1.4.0->accelerate==0.18.0->-r requirements.txt (line 5)) (1.2.1)
Installing collected packages: safetensors, transformers, accelerate, datasets
  Attempting uninstall: safetensors
    Found existing installation: safetensors 0.3.1
    Uninstalling safetensors-0.3.1:
      Successfully uninstalled safetensors-0.3.1
  Attempting uninstall: transformers
    Found existing installation: transformers 4.29.1
    Uninstalling transformers-4.29.1:
      Successfully uninstalled transformers-4.29.1
  Attempting uninstall: accelerate
    Found existing installation: accelerate 0.19.0
    Uninstalling accelerate-0.19.0:
      Successfully uninstalled accelerate-0.19.0
  Attempting uninstall: datasets
    Found existing installation: datasets 2.12.0
    Uninstalling datasets-2.12.0:
      Successfully uninstalled datasets-2.12.0
Successfully installed accelerate-0.18.0 datasets-2.10.1 safetensors-0.3.0 transformers-4.28.0
Processing g:\git-jv\ai\oobabooga_one-click-installersk-installers\text-generation-webui\repositories\gptq-for-llama
  Preparing metadata (setup.py): started
  Preparing metadata (setup.py): finished with status 'error'
  error: subprocess-exited-with-error

  python setup.py egg_info did not run successfully.
  exit code: 1

  [8 lines of output]
  Traceback (most recent call last):
    File "<string>", line 2, in <module>
    File "<pip-setuptools-caller>", line 34, in <module>
    File "G:\git-jv\ai\oobabooga_one-click-installersk-installers\text-generation-webui\repositories\GPTQ-for-LLaMa\setup.py", line 2, in <module>
      from torch.utils import cpp_extension
    File "G:\git-jv\ai\oobabooga_one-click-installersk-installers\installer_files\env\lib\site-packages\torch\__init__.py", line 91, in <module>
      cuda_version_1 = cuda_version.replace('.', '_')
  AttributeError: 'float' object has no attribute 'replace'
  [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
error: metadata-generation-failed

Encountered error while generating package metadata.

See above for output.

note: This is an issue with the package mentioned above, not pip.
hint: See above for details.
```

JulianVolodia committed:
Casting to str of cuda_version as it is not used once

janeyx99 commented on torch/__init__.py:
```diff
@@ -88,7 +88,7 @@ def _running_with_deploy():
     import glob
     if cuda_version and all(not glob.glob(os.path.join(p, 'cudart64*.dll')) for p in dll_paths):
-        cuda_version_1 = cuda_version.replace('.', '_')
+        cuda_version_1 = str(cuda_version).replace('.', '_')
```
Maybe better to make cuda_version a string before line 90?

JulianVolodia replied:
actually, this is idiomatic, if cuda_version is None --> if cuda_version would not be True if cuda_version is None

so I did object conversion under cuda_version evaluation to True condition


actually, this is draft - but look for some domain knowledge if 12.1 should be supported experimentally and under-the-table only, or from nightly indeed

JulianVolodia committed:
Revert "Wrapping cuda_version with str() in case it is float type"

This reverts commit 749a2bb3409ecf8d577c0be3df4eb065a0956515.

JulianVolodia commented:
result now is sth like that:

```
    File "G:\git-jv\ai\oobabooga_one-click-installersk-installers\installer_files\env\lib\site-packages\setuptools\_distutils\command\build_ext.py", line 346, in run
      self.build_extensions()
    File "G:\git-jv\ai\oobabooga_one-click-installersk-installers\installer_files\env\lib\site-packages\torch\utils\cpp_extension.py", line 500, in build_extensions
      _check_cuda_version(compiler_name, compiler_version)
    File "G:\git-jv\ai\oobabooga_one-click-installersk-installers\installer_files\env\lib\site-packages\torch\utils\cpp_extension.py", line 382, in _check_cuda_version
      torch_cuda_version = packaging.version.parse(torch.version.cuda)
    File "G:\git-jv\ai\oobabooga_one-click-installersk-installers\installer_files\env\lib\site-packages\pkg_resources\_vendor\packaging\version.py", line 49, in parse
      return Version(version)
    File "G:\git-jv\ai\oobabooga_one-click-installersk-installers\installer_files\env\lib\site-packages\pkg_resources\_vendor\packaging\version.py", line 264, in __init__
      match = self._regex.search(version)
  TypeError: expected string or bytes-like object
  [end of output]

  note: This error originates from a subprocess, and is likely not a problem with pip.
  Rolling back uninstall of quant-cuda
```

JulianVolodia committed:
Dumb add of CUDA 12.1 minimal clang & gcc version requirements

JulianVolodia committed:
Add stub of CUDA 12.1 minimal clang & gcc version requirements

JulianVolodia commented:
in nightly branch, torch/utils/cpp_extension.py I also do not see clang nor gcc minimum version specified 

I added them to check what would happen.

Try to workaround nightly problem with this end with list index out of range in _get_cuda_arch_flags

```

    File "G:\git-jv\ai\oobabooga_one-click-installersk-installers\installer_files\env\lib\site-packages\setuptools\_distutils\dist.py", line 988, in run_command
      cmd_obj.run()
    File "G:\git-jv\ai\oobabooga_one-click-installersk-installers\installer_files\env\lib\site-packages\setuptools\command\build_ext.py", line 84, in run
      _build_ext.run(self)
    File "G:\git-jv\ai\oobabooga_one-click-installersk-installers\installer_files\env\lib\site-packages\setuptools\_distutils\command\build_ext.py", line 346, in run
      self.build_extensions()
    File "G:\git-jv\ai\oobabooga_one-click-installersk-installers\installer_files\env\lib\site-packages\torch\utils\cpp_extension.py", line 848, in build_extensions
      build_ext.build_extensions(self)
    File "G:\git-jv\ai\oobabooga_one-click-installersk-installers\installer_files\env\lib\site-packages\setuptools\_distutils\command\build_ext.py", line 468, in build_extensions
      self._build_extensions_serial()
    File "G:\git-jv\ai\oobabooga_one-click-installersk-installers\installer_files\env\lib\site-packages\setuptools\_distutils\command\build_ext.py", line 494, in _build_extensions_serial
      self.build_extension(ext)
    File "G:\git-jv\ai\oobabooga_one-click-installersk-installers\installer_files\env\lib\site-packages\setuptools\command\build_ext.py", line 246, in build_extension
      _build_ext.build_extension(self, ext)
    File "G:\git-jv\ai\oobabooga_one-click-installersk-installers\installer_files\env\lib\site-packages\setuptools\_distutils\command\build_ext.py", line 549, in build_extension
      objects = self.compiler.compile(
    File "G:\git-jv\ai\oobabooga_one-click-installersk-installers\installer_files\env\lib\site-packages\torch\utils\cpp_extension.py", line 808, in win_wrap_ninja_compile
      cuda_post_cflags = win_cuda_flags(cuda_post_cflags)
    File "G:\git-jv\ai\oobabooga_one-click-installersk-installers\installer_files\env\lib\site-packages\torch\utils\cpp_extension.py", line 678, in win_cuda_flags
      cflags + _get_cuda_arch_flags(cflags))
    File "G:\git-jv\ai\oobabooga_one-click-installersk-installers\installer_files\env\lib\site-packages\torch\utils\cpp_extension.py", line 1782, in _get_cuda_arch_flags
      arch_list[-1] += '+PTX'
  IndexError: list index out of range
```

JulianVolodia commented:
Deepest sorry for making some loud here, if you close it as unneeded for latest versions troubleshooting, I would completely understand.

JulianVolodia commented:
I found in failing that linter warning blocks only linux job

https://github.com/pytorch/pytorch/actions/runs/5017918767/jobs/8997840985?pr=101806#step:10:1350

janeyx99 reviewed (commented):
In general, we want all jobs to be green before merging. The lintrunner job logs say to add spaces before comments, which is not a hard fix.

janeyx99 reviewed (commented):
In general, we want all jobs to be green before merging. The lintrunner job logs say to add spaces before comments, which is not a hard fix.

janeyx99 commented:
Looping in @atalman about 12.1 changes in cpp_extensions. I'm not sure whether these changes should be made.

JulianVolodia commented:
> Looping in @atalman about 12.1 changes in cpp_extensions. I'm not sure whether these changes should be made.

Sorry for messing, I am pretty sure those shouldn't.

JulianVolodia commented:
I added suggested change. Added extra spaces before the hash starting the python's comment.

Fail was caused by linter's warning in late-stage but only in one pipeline. Tell me if you would like to add some linter check, to fail early and not involve pipelines work until code is passing linter check. I could do it.

JulianVolodia committed:
Revert "Casting to str of cuda_version as it is not used once"

This reverts commit 88bdc81f46e501bde6e85a92e6e9cedd779bd85e.

JulianVolodia committed:
Revert "Dumb add of CUDA 12.1 minimal clang & gcc version requirements"

This reverts commit 71504ba7f5532bc82744d81673c305183c092405.

JulianVolodia commented:
reverted pointless commits, and fixed git fork history after force push. I will reopen, and if I could join you working on testing new versions compatible to 12.1 I would be proud of doing so.

malfet commented:
Thank you for the update. Two comments:
 - Please remove `not tested` comment, as it is tested by CI
 - Please update PR description and reference where it doc that specifies minimum and maximum supported clang versions (I.e. I don't see why cuda-12 would not support clang-15)

atalman commented:
@JulianVolodia if you trying to fix this issue:  https://github.com/pytorch/pytorch/issues/101801 please take a look at this comment: https://github.com/pytorch/pytorch/issues/101801#issuecomment-1563545421

JulianVolodia commented:
Thanks @atalman for comment here and in issue thread.

Thanks @malfet - you are right. I do not have reference for this. It was (originally, now the commit message is different) "dumb" fix for errors I've seen when trying to integrate nightly with CUDA support, but turns out - I have been using wrong version.

I will close this, as clearly that is not pytorch issue.
